src/bin/pg_autoctl/pgsql.c

```diff
@@ -230,32 +428,83 @@ static PGconn *
 pgsql_retry_open_connection(PGSQL *pgsql)
 {
 	PGconn *connection = NULL;
-	int attempts = 0;
-	bool retry = true;
 	bool connectionOk = false;
+
 	uint64_t startTime = time(NULL);
+	PGPing lastWarningMessage = PQPING_OK;
+	uint64_t lastWarningTime = 0;
 
 	log_warn("Failed to connect to \"%s\", retrying until "
 			 "the server is ready", pgsql->connectionString);
 
-	while (retry)
+	/* should not happen */
+	if (pgsql->retryPolicy.maxR == 0)
+	{
+		return NULL;
+	}
+
+	/* reset our internal counter before entering the retry loop */
+	pgsql->retryPolicy.attempts = 0;
+
+	while (!connectionOk)
 	{
+		int sleep = 0;
 		uint64_t now = time(NULL);
 
-		if ((now - startTime) >= POSTGRES_PING_RETRY_TIMEOUT)
+		/* Any signal is reason enough to break out from this retry loop. */
+		if (asked_to_quit ||
+			asked_to_stop ||
+			asked_to_stop_fast ||
+			asked_to_reload)
 		{
-			log_warn("Failed to connect to \"%s\" after %d attempts, "
-					 "stopping now", pgsql->connectionString, attempts);
 			break;
 		}
 
-		++attempts;
+		/*
+		 * We stop retrying as soon as we have spent all of our time budget or
+		 * all of our attempts count budget, whichever comes first.
+		 *
+		 * maxR = 0 (zero) means no retry at all, checked before the loop
+		 * maxR < 0 (zero) means unlimited number of retries
+		 */
+		if ((now - startTime) >= pgsql->retryPolicy.maxT ||
+			(pgsql->retryPolicy.maxR > 0 &&
+			 pgsql->retryPolicy.attempts >= pgsql->retryPolicy.maxR))
+		{
+			log_warn("Failed to connect to \"%s\" "
```

Comment:
I think this should be a `log_error` instead. 

Also, this error doesn't give any indication what the problem with connecting is. The previous error that contained all this information:
```
11:57:14 18700 ERROR Connection to database failed: could not connect to server: Connection refused
11:57:14 18700 ERROR 	Is the server running on host "172.24.28.118" and accepting
11:57:14 18700 ERROR 	TCP/IP connections on port 5432?
11:57:14 18700 ERROR Failed to connect to monitor database at "postgres://autoctl_node@172.24.28.118:5432/pg_auto_failover?sslmode=require", see above for details
```

Can we add this to the final fail in some way?

Reply:
Added now, thanks!